### DISCUSSION TITLE
Support wildcard as argument for allowing all attributes/columns

### DISCUSSION THREAD

**Discussion by marknuyens** [spatie/laravel-query-builder#609] [Ideas]
I believe it would be a nice addition if you weren't bound to the attributes/columns that you need to specify, if you would like to include all columns to be allowed. This could result in a single additional argument check for an string asterix, for example.

Anyway, hope this is something others may find useful as well.

Thanks.

**Comment by vincent-lu** [comment]
Would really love having this. In my case I'd prefer to have `*` to indicate "all (columns)". Does this have something to do with Laravel not knowing what columns are availabe in a table?

**Reply by marknuyens** [reply]
I don’t believe so; currently, I am accepting the requested query parameter (e.g. “filter”) as the value to allow for this feature in the meantime. Like so:

`$this->allowedFilters(array_keys(request(‘filter’) ?? []))`;

It feels a bit hacky and more importantly, repetitive.

One argument I could think of for not supporting this behavior is security related, where users could specify unanticipated input. The current solution (i.e. by specifying allowed columns), would prevent this.

**Reply by vincent-lu** [reply]
Yeah I thought about that but end up adding a function like this in my models where I know it's a run-of-a-mill type model:

```php
public static function getAllFields(): array
{
    $model = new static();

    return array_merge(['id'], $model->getFillable());
}
```

then in the controller I just call
```php
->allowedFields(Product::getAllFields())
```

I get the security argument, but it really depends on use case; for example in my situation all access are protected by auth gate and only company stuff can access, and the model in question doesn't have particularly sensitive information. If one can use `protected $guarded = []` in models, I don't think why a wildcard `*` is less kosher as long as the dev understands.

**Reply by marknuyens** [reply]
I was actually considering a similar approach, but using different properties on the model. For example:

`$filterable = […];
$sortable=[…];
Etc…`

And then pretty much use your way of obtaining them. 😎

**Comment by freekmurze** [comment]
This is coming in the next major version (v7) of the package.

You will be able to use a wildcard `"*"` to allow all filters, sorts, and includes:

```php
QueryBuilder::for(User::class)
    ->allowedFilters("*")
    ->allowedSorts("*")
    ->allowedIncludes("*");
```

When using the wildcard, all requested filters/sorts/includes from the query string will be allowed without needing to specify each one explicitly.

Docs will be available at https://spatie.be/docs/laravel-query-builder/v7/features/filtering once v7 is released.
